### PR TITLE
fix: Cache to check if a data element has a link with rule variable [TECH-576]

### PR DESCRIPTION
**Pull request by enricocolasante**

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/update/postprocess/PublishEventPostProcessor.java [hunk 1]
```diff
@@ -43,26 +41,7 @@
...
-
-                if ( isLinkedWithRuleVariable )
-                {
-                    break;
-                }
-            }
-        }
-
-        if ( !ctx.getImportOptions().isSkipNotifications() && isLinkedWithRuleVariable )
+        if ( !ctx.getImportOptions().isSkipNotifications() )
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/update/postprocess/PublishEventPostProcessor.java:44]:
```suggestion
        boolean isLinkedWithRuleVariable = false;

        for ( final DataValue dv : event.getDataValues() )
        {
            final DataElement dataElement = ctx.getDataElementMap().get( dv.getDataElement() );

            if ( dataElement != null )
            {
                // TODO: luciano preload the value
                isLinkedWithRuleVariable = ctx.getServiceDelegator().getProgramRuleVariableService()
                    .isLinkedToProgramRuleVariable( ctx.getProgramsMap().get( event.getProgram() ), dataElement );

                if ( isLinkedWithRuleVariable )
                {
                    break;
                }
            }
        }

        if ( !ctx.getImportOptions().isSkipNotifications() && isLinkedWithRuleVariable )

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/update/postprocess/PublishEventPostProcessor.java [hunk 2]
```diff
@@ -27,8 +27,6 @@
  */
 package org.hisp.dhis.dxf2.events.importer.update.postprocess;
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/update/postprocess/PublishEventPostProcessor.java:29]:
```suggestion
import org.hisp.dhis.dataelement.DataElement;
import org.hisp.dhis.dxf2.events.event.DataValue;
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/update/postprocess/PublishEventPostProcessor.java [hunk 3]
```diff
@@ -26,7 +26,6 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 package org.hisp.dhis.dxf2.events.importer.update.postprocess;
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/update/postprocess/PublishEventPostProcessor.java:28]:
```suggestion
package org.hisp.dhis.dxf2.events.importer.update.postprocess;

```

**Review by Bekkalizer — APPROVED** [review]
Looks good to me!

**Review by zubaira — APPROVED** [review]

**Comment by larshelge** [comment]
Great @enricocolasante , I think this looks good.

File: dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/DefaultProgramRuleVariableService.java [hunk 4]
```diff
@@ -65,6 +71,7 @@ public DefaultProgramRuleVariableService( ProgramRuleVariableStore programRuleVa
     public long addProgramRuleVariable( ProgramRuleVariable programRuleVariable )
     {
         programRuleVariableStore.save( programRuleVariable );
+        programRuleVariablesCache.invalidateAll();
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/DefaultProgramRuleVariableService.java:74]: This is okay, though I don't think this makes it solid as we use the metadata importer to import all metadata including ProgramRuleVariables, which means the cache won't be cleared mostly. This then gives a "false sense of security". 

Could we maybe instead just introduce a `isLinkedToProgramRuleVariableCached` method where it is explicit that the caller is getting a cached result for now? Or just remove the invalidation call and add the caching behavior in the Java docs of this method?

Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/DefaultProgramRuleVariableService.java:74]: The method is used only in the postprocessor, so I just renamed the method.
And I maintained the invalidation in the delete method that is called in the deletion handlers

**Review by larshelge — APPROVED** [review]

**Pull request merged**